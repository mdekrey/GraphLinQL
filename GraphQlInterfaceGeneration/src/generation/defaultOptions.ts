import { Options } from "./Options";

export const defaultOptions: Options = {
  namespace: "GraphLinqQL.Interfaces",
  useNullabilityIndicator: true,
  scalarTypes: {
    Int: { csharpNullable: false, csharpType: "int" },
    Float: { csharpNullable: false, csharpType: "double" },
    String: { csharpNullable: true, csharpType: "string" },
    Boolean: { csharpNullable: false, csharpType: "bool" }
  },
  using: [
    "GraphLinqQL",
    "GraphLinqQL.Introspection",
    "System",
    "System.Collections",
    "System.Collections.Generic",
    "System.Collections.Immutable",
    "System.Linq"
  ],
  deserializer: "System.Text.Json.JsonSerializer.Deserialize",
  introspection: false
};
