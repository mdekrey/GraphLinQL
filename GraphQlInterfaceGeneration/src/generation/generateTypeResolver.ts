import { Options } from "./Options";
import { GraphQLSchema, GraphQLInputObjectType, isInputObjectType } from "graphql";
import { getInputTypeName } from "./getInputTypeName";

export function generateTypeResolver(schema: GraphQLSchema, options: Options) {
  const inputObjectTypes = Object.keys(schema.getTypeMap())
    .map(typeName => schema.getType(typeName))
    .filter((t): t is GraphQLInputObjectType => (t ? isInputObjectType(t) : false));
  const hasId = Boolean(options.scalarTypes["ID"]);
  return `
public class TypeResolver : IGraphQlTypeResolver
{
    public Type Resolve(string name)
    {
        switch (name)
        {
            ${hasId ? "" : scalarType("ID", "String")}
            ${Object.keys(options.scalarTypes).map(t => scalarType(t, t)).join(`
            `)}
            ${inputObjectTypes.map(
              input => `case "${input.name}":
                return typeof(${getInputTypeName(input, options, false)});`
            ).join(`
            `)}
            default:
                throw new ArgumentException("Unknown type " + name, nameof(name));
        }
    }
}`;
  function scalarType(label: string, t: string) {
    return `case "${label}":
                return typeof(${`${options.scalarTypes[t].csharpType}${
                  options.scalarTypes[t].csharpNullable ? "" : "?"
                }`});`;
  }
}
