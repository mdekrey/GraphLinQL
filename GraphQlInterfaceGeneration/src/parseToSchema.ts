import { parse, buildASTSchema, GraphQLSchema } from "graphql";

export function parseToSchema(graphQlSchema: string): GraphQLSchema {
  const astDocument = parse(graphQlSchema);

  const backcompatOptions = { commentDescriptions: true };
  const schema = buildASTSchema(astDocument, backcompatOptions);

  return schema;
}
