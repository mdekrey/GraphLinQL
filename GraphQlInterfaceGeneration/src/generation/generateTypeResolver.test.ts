import { parseToSchema } from "../parseToSchema";
import { getStarWarsGraphQlSchemaV3 } from "../resources.test";
import { defaultOptions } from "./defaultOptions";
import { generateTypeResolver } from "./generateTypeResolver";

it("can generate v3 types", async function() {
  const schema = await getSampleSchema(getStarWarsGraphQlSchemaV3);
  expect(generateTypeResolver(schema, defaultOptions)).toMatchSnapshot();
});

async function getSampleSchema(getSchema: () => Promise<string>) {
  const schema = parseToSchema(await getSchema());
  return schema;
}
