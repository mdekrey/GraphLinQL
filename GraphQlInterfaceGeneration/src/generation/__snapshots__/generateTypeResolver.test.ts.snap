// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can generate v3 types 1`] = `
"
public class TypeResolver : IGraphQlTypeResolver
{
    public Type Resolve(string name)
    {
        switch (name)
        {
            case \\"ID\\":
                return typeof(string);
            case \\"Int\\":
                return typeof(int?);
            case \\"Float\\":
                return typeof(double?);
            case \\"String\\":
                return typeof(string);
            case \\"Boolean\\":
                return typeof(bool?);
            case \\"ReviewInput\\":
                return typeof(ReviewInput);
            case \\"ColorInput\\":
                return typeof(ColorInput);
            default:
                throw new ArgumentException(\\"Unknown type \\" + name, nameof(name));
        }
    }
}"
`;
